File: app/src/main/java/no/nav/tilbakemeldingsmottak/consumer/email/aad/AADMailConfiguration.java
package no.nav.tilbakemeldingsmottak.consumer.email.aad;

import com.azure.core.http.HttpClient;
import com.azure.core.util.HttpClientOptions;
import com.azure.identity.ClientSecretCredentialBuilder;
import com.microsoft.graph.authentication.TokenCredentialAuthProvider;
import com.microsoft.graph.httpcore.HttpClients;
import com.microsoft.graph.requests.GraphServiceClient;
import lombok.RequiredArgsConstructor;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.context.annotation.Profile;

@Configuration
@RequiredArgsConstructor
@Profile("nais")
public class AADMailConfiguration {

    private final AADProperties aadProperties;

    @Bean
    GraphServiceClient getGraphClient() {

        final TokenCredentialAuthProvider authenticationProvider = getTokenProvider();

        final var graphHttpClient =
                HttpClients.createDefault(authenticationProvider)
                        .newBuilder()
                        .build();

        return GraphServiceClient
                .builder()
                .httpClient(graphHttpClient)
                .buildClient();
    }

    private TokenCredentialAuthProvider getTokenProvider() {
        HttpClientOptions clientOptions = new HttpClientOptions();

        HttpClient azHttpClient = HttpClient.createDefault(clientOptions);

        var clientSecretCredential = new ClientSecretCredentialBuilder()
                .clientId(aadProperties.getClientId())
                .clientSecret(aadProperties.getClientSecret())
                .tenantId(aadProperties.getTenant())
                .httpClient(azHttpClient)
                .build();

        return new TokenCredentialAuthProvider(clientSecretCredential);
    }
}

File: app/src/main/kotlin/no/nav/tilbakemeldingsmottak/consumer/email/aad/AADMailConfiguration.kt
package no.nav.tilbakemeldingsmottak.consumer.email.aad

import com.azure.core.credential.TokenCredential
import com.azure.identity.ClientSecretCredentialBuilder
import com.microsoft.graph.serviceclient.GraphServiceClient
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.context.annotation.Profile

@Configuration
@Profile("nais")
class AADMailConfiguration(private val aadProperties: AADProperties) {

    private val MICROSOFT_GRAPH_SCOPES: Set<String> = java.util.Set.of(
        "Mail.Send"
    )

    @Bean
    fun getGraphClient(): GraphServiceClient {

        return GraphServiceClient(getTokenCredentials(), *MICROSOFT_GRAPH_SCOPES.toTypedArray())
    }

    private fun getTokenCredentials(): TokenCredential {
        return ClientSecretCredentialBuilder()
            .tenantId(aadProperties.tenant)
            .clientId(aadProperties.clientId)
            .clientSecret(aadProperties.clientSecret)
            .build()
    }
}
